### DISCUSSION TITLE
lyd_find_xpath alternative results

### DISCUSSION THREAD

**Discussion by choppsv1** [CESNET/libyang#2148] [Ideas]
Currently lyd_find_xpath* returns a set of nodes. These nodes are references to the tree that is passed in (either from the `ctx` or directly in variant 3 and 4).

I'd really like to use this function to actually trim the passed in tree, as for me the entire tree represent the results (e.g., I'm passing them back to a client in XML format so they are from the root)

I do not actually need the `LY_SET` of all the matching leaf nodes, but rather the tree w/o any non-matching nodes in it. Additionally this could be operating on a large set of data where the matches might count in the millions so it's very inefficient to allocate and fill the unneeded LY_SET.

Would it be possible to be provided with an analogous API functions that do this? e.g.,

```
...
LY_ERR lyd_trim_xpath3(struct lyd_node *tree, const char *xpath, struct lyxp_var *vars);
...
```

name-wise trim, filter, whatever seems best :)

**Comment by michalvasko** [comment]
Alright, this can be added and hopefully I have come up with a reasonable algorithm. My only question is what parameters do you need? The required ones are only `tree` and `xpath`, all others (look at `lyd_find_xpath4()`) are optional for more specific use-cases.

**Comment by choppsv1** [comment]
For our current mark/sweep we use `lyd_find_xpath3()`, and we don't currently use xpath variables; however, this could change in the future.

We use JSON prefix style (module names) in our xpath strings (these are actually from the user e.g., a CLI command) so that's why we don't need `lyd_find_xpath4()` variant, but maybe other users might use the other prefix types? I don't know what prefix_data is for. :)

So I think we'd use (tree, xpath, vars), but maybe others might want the full (tree, xpath, format, vars)?

**Comment by michalvasko** [comment]
Okay, the function is implemented and should even be quite efficient. If the need for more parameters arises in future, a new variant can be added.

**Comment by choppsv1** [comment]
Thanks!
